File: webhooks/go/verify_test.go
package fletchwebhook

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"
)

type vector struct {
	Name             string `json:"name"`
	Secret           string `json:"secret"`
	Header           string `json:"header"`
	Body             string `json:"body"`
	Now              int64  `json:"now"`
	ToleranceSeconds int64  `json:"toleranceSeconds"`
	Valid            bool   `json:"valid"`
	Reason           string `json:"reason"`
}

func loadVectors(t *testing.T) []vector {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("..", "fixtures", "vectors.json"))
	if err != nil {
		t.Fatal(err)
	}
	var file struct {
		Vectors []vector `json:"vectors"`
	}
	if err := json.Unmarshal(raw, &file); err != nil {
		t.Fatal(err)
	}
	return file.Vectors
}

func TestVectors(t *testing.T) {
	for _, v := range loadVectors(t) {
		t.Run(v.Name, func(t *testing.T) {
			got := Verify(v.Secret, v.Header, []byte(v.Body), time.Duration(v.ToleranceSeconds)*time.Second, time.Unix(v.Now, 0))
			if got != v.Valid {
				t.Fatalf("got %v, want %v: %s", got, v.Valid, v.Reason)
			}
		})
	}
}

func TestSignRoundTrip(t *testing.T) {
	header := SignWebhookBody("whsec_test", []byte(`{"id":"dl_1"}`), 1789200000)
	if !Verify("whsec_test", header, []byte(`{"id":"dl_1"}`), 5*time.Minute, time.Unix(1789200010, 0)) {
		t.Fatal("own signature did not verify")
	}
	if Verify("whsec_test", header, []byte(`{"id":"dl_2"}`), 5*time.Minute, time.Unix(1789200010, 0)) {
		t.Fatal("a different body verified")
	}
}

File: webhooks/typescript/verify.test.ts
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { test } from "node:test";
import { signWebhookBody, verifyFletchSignature } from "./verify.ts";

interface Vector {
  name: string;
  secret: string;
  header: string;
  body: string;
  now: number;
  toleranceSeconds: number;
  valid: boolean;
  reason: string;
}

const vectors = JSON.parse(readFileSync(new URL("../fixtures/vectors.json", import.meta.url), "utf8")) as { vectors: Vector[] };

for (const vector of vectors.vectors) {
  test(`${vector.valid ? "accepts" : "rejects"}: ${vector.name}`, function run() {
    const result = verifyFletchSignature(vector.secret, vector.header, vector.body, { now: vector.now, toleranceSeconds: vector.toleranceSeconds });
    assert.equal(result, vector.valid, vector.reason);
  });
}

test("signs the way Fletch does", function run() {
  const header = signWebhookBody("whsec_test", '{"id":"dl_1"}', 1789200000);
  assert.ok(verifyFletchSignature("whsec_test", header, '{"id":"dl_1"}', { now: 1789200010 }));
  assert.ok(!verifyFletchSignature("whsec_test", header, '{"id":"dl_2"}', { now: 1789200010 }));
});

test("signs the raw bytes, not a re-encoded string", function run() {
  const body = new TextEncoder().encode('{"symbol":"AAPL","note":"€"}');
  const header = signWebhookBody("whsec_test", body, 1789200000);
  assert.ok(verifyFletchSignature("whsec_test", header, body, { now: 1789200000 }));
  assert.ok(verifyFletchSignature("whsec_test", header, '{"symbol":"AAPL","note":"€"}', { now: 1789200000 }));
});
